src/Core/MySQLReplication.cpp

```diff
@@ -0,0 +1,826 @@
+        {
+            std::cerr << "Row[" << i << "]: " << applyVisitor(to_string, rows[i]) << std::endl;
+        }
+    }
+
+    void DryRunEvent::parseImpl(ReadBuffer & payload) { payload.ignore(header.event_size - EVENT_HEADER_LENGTH); }
+
+    void DryRunEvent::dump() const
+    {
+        header.dump();
+        std::cerr << "[DryRun Event]" << std::endl;
+    }
+
+    void MySQLFlavor::readPayloadImpl(ReadBuffer & payload)
+    {
+        UInt16 header = static_cast<unsigned char>(*payload.position());
+        switch (header)
+        {
+            case PACKET_EOF:
+                throw ReplicationError("Master maybe lost", ErrorCodes::UNKNOWN_EXCEPTION);
+            case PACKET_ERR:
+                ERR_Packet err;
+                err.readPayloadImpl(payload);
+                throw ReplicationError(err.error_message, ErrorCodes::UNKNOWN_EXCEPTION);
+        }
+        // skip the header flag.
+        payload.ignore(1);
+
+        EventType event_type = static_cast<EventType>(*(payload.position() + 4));
+        switch (event_type)
+        {
+            case FORMAT_DESCRIPTION_EVENT: {
+                event = std::make_shared<FormatDescriptionEvent>();
+                event->parseHeader(payload);
+                event->parseEvent(payload);
+                position.updateLogPos(event->header.log_pos);
+                break;
+            }
+            case ROTATE_EVENT: {
+                event = std::make_shared<RotateEvent>();
+                event->parseHeader(payload);
+                event->parseEvent(payload);
+
+                auto rotate = std::static_pointer_cast<RotateEvent>(event);
+                position.updateLogPos(event->header.log_pos);
+                position.updateLogName(rotate->next_binlog);
+                break;
+            }
+            case QUERY_EVENT: {
+                event = std::make_shared<QueryEvent>();
+                event->parseHeader(payload);
+                event->parseEvent(payload);
+
+                auto query = std::static_pointer_cast<QueryEvent>(event);
+                switch (query->typ)
+                {
+                    case BEGIN:
+                    case XA: {
+                        event = std::make_shared<DryRunEvent>();
+                        break;
+                    }
+                    default:
+                        position.updateLogPos(event->header.log_pos);
+                }
+
+                break;
+            }
+            case XID_EVENT: {
+                event = std::make_shared<XIDEvent>();
+                event->parseHeader(payload);
+                event->parseEvent(payload);
+                position.updateLogPos(event->header.log_pos);
+                break;
+            }
+            case TABLE_MAP_EVENT: {
+                event = std::make_shared<TableMapEvent>();
+                event->parseHeader(payload);
+                event->parseEvent(payload);
+                position.updateLogPos(event->header.log_pos);
```

Comment:
If we just sync to this event and shutdown, next time server restart the table_map event will be lost.
We don't need to update the position here.